libafl/src/feedbacks/stdio.rs

```diff
@@ -0,0 +1,210 @@
+//! Feedback and metatadata for stderr and stdout.
+
+use alloc::string::{String, ToString};
+
+use libafl_bolts::{impl_serdeany, Named};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    corpus::Testcase,
+    events::EventFirer,
+    executors::ExitKind,
+    feedbacks::Feedback,
+    observers::{ObserversTuple, StdErrObserver, StdOutObserver},
+    state::State,
+    Error, HasMetadata,
+};
+
+/// Metadata for `StdOutFeddback`,
+#[derive(Debug, Serialize, Deserialize)]
+pub struct StdOutMetadata {
+    #[allow(missing_docs)]
+    pub stdout: String,
+}
+
+impl_serdeany!(StdOutMetadata);
+
+/// Nop feedback that annotates stdout in the new testcase. The testcase
+/// is never interesting (use with an OR).
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct StdOutFeedback {
+    name: String,
+}
+
+impl<S> Feedback<S> for StdOutFeedback
+where
+    S: State,
+{
+    #[allow(clippy::wrong_self_convention)]
+    fn is_interesting<EM, OT>(
+        &mut self,
+        _state: &mut S,
+        _manager: &mut EM,
+        _input: &S::Input,
+        _observers: &OT,
+        _exit_kind: &ExitKind,
+    ) -> Result<bool, Error>
+    where
+        EM: EventFirer<State = S>,
+        OT: ObserversTuple<S>,
+    {
+        Ok(false)
+    }
+
+    /// Append to the testcase the generated metadata in case of a new corpus item.
+    #[inline]
+    fn append_metadata<EM, OT>(
+        &mut self,
+        _state: &mut S,
+        _manager: &mut EM,
+        observers: &OT,
+        testcase: &mut Testcase<S::Input>,
+    ) -> Result<(), Error>
+    where
+        OT: ObserversTuple<S>,
+        EM: EventFirer<State = S>,
+    {
+        let observer = observers
+            .match_name::<StdOutObserver>(self.name())
+            .ok_or(Error::illegal_state("StdOutObserver is missing"))?;
+        let buffer = observer
+            .stdout
+            .as_ref()
+            .ok_or(Error::illegal_state("StdOutObserver has no stdout"))?;
+        let stdout = String::from_utf8_lossy(buffer).into_owned();
+
+        testcase
+            .metadata_map_mut()
+            .insert(StdOutMetadata { stdout });
+
+        Ok(())
+    }
+
+    /// Discard the stored metadata in case that the testcase is not added to the corpus.
+    #[inline]
+    fn discard_metadata(&mut self, _state: &mut S, _input: &S::Input) -> Result<(), Error> {
+        Ok(())
+    }
+}
+
+impl Named for StdOutFeedback {
+    #[inline]
+    fn name(&self) -> &str {
+        self.name.as_str()
+    }
+}
+
+impl StdOutFeedback {
+    /// Creates a new [`StdOutFeedback`].
```

Comment:
I would either document that the name is used to look up the observer, or leave out this constructor completely